### ISSUE TITLE
Cythonize and parallelize the core of the spatial random field generation (RandMeth)

### ISSUE THREAD

**Issue by LSchueler** [GeoStat-Framework/GSTools#16]
I've replaced the vectorized numpy-based randomization method (and hard to read because of all the chunking and broadcasting) with a Cython implementation, which utilizes OpenMP for parallel execution of the main loop.

First tests have shown a nearly linear decrease of runtime with the number of cores (up to 4).
And numpy's incredible appetite for memory is completely circumvented. Now it is possible to generate huge 3d fields in one step, without gluing the fields together.

But before merging this into the master branch, there are a few things to clarify:

- Before we used deployment tools, it made sense to assume that not everybody has Cython or a C compiler at hand, thus we supported pure Python implementations alongside the Cython code. Should we drop this assumption and remove the pure Python implementations? This would make all the broadcasting much easer.

- Can the deployment tools compile against OpenMP?

- Should the IncomprRandMeth also be implemented in Cython or is it not worth the effort?

- I'm pretty sure, that the current setup.py and the Cython code would fall back to a non-parallelized version, if no OpenMP support is found, but this has to be checked.

See the `cythonize_randmeth` branch for the implementation, with the current commit being 313019aa.

**Comment by MuellerSeb** [comment]
This is awesome! Thanks for the nice work, that's a neat improvement.

I think the Cython implementation of the RandMeth method is just fine for the package. I think we should create a subfolder in ``field`` for cython files, for future implementations (for example the ``IncomprRandMeth``, do you think, it's that much work?)

But I got some issues with the OpenMP approach, even though it brings a huge benefit!

As you assumed, the deployment is a problem here. I think it should be possible to build wheels agains OpenMP on Travis and Appveyor, but it's not guaranteed, that the user provides OpenMP in the end.. So this creates a problematic scenario, where people install the package with pip without an error, but can't use it in the end. I don't know if we can build multiple wheels (with/without OpenMP) and use a pip-flag to choose?! A short google summary didn't bring me to a solution...

At the moment, I would suggest a build-flag for the setup.py, so that a user with interest in OpenMP could build the package on his/her own, we just need to provide a guide on that. Maybe the variogram estimator could run in parallel as well then?

About the python implementations: I thought, that the python implementations are just beside the cython implementation and we use a try-except block for importing, like with the variogram estimator? Where does the broadcasting get easier? Maybe I'm a bit naive here...

**Comment by MuellerSeb** [comment]
Maybe we could provide a flag for ``pip install`` with this:
https://stackoverflow.com/a/46582365/6696397

Then, the sequential cython implementation would be the standard way of installation via wheels and we can provide OpenMP support via:

    pip install gstools --install-option='--openmp'

So in this case the package will be built locally and the user has to provide cython, openmp and a c-compiler. With this approach we also don't have to check for OpenMP since it's in the users hand to care for that.

@LSchueler : what do you think about that?

**Comment by MuellerSeb** [comment]
Also on that topic:
https://stackoverflow.com/q/677577/6696397

**Comment by Hugovdberg** [comment]
Why did you choose `cython` over `numba`? It appears to me the numba JIT compiler would be suitable for this code, and includes the `LLVM` compiler that supports parallelisation without external user installed dependencies. Of course `numba` isn't the most lightweight dependency, so you might want to add it with an install flag.

**Comment by LSchueler** [comment]
@Hugovdberg Our first use of Cython was for the variogram estimation. The main reason was actually not speed or memory, but readability and maintainability. Some algorithms are not very well suited for the broadcasting style of Numpy.
We did that quite some time ago, when Numba was a lot less mature than it is now. And many of the big scientific Python packages rely on Cython, which makes it some kind of de facto standard in our community. Some examples are pandas, scipy (including numpy), scikit-learn, and mpi4py.
Cython also has a much larger support base for both, finances and development, in contrast to numba, which is mainly supported by Anaconda.
And as far as I know, deployment of Cython code seems a lot easier than Numba code. - I admit, that adding the OpenMP dependency contradicts this statement to some extent. ;-)
But for now, installing GSTools via pip is working pretty great on all commonly used platforms.

Keep in mind, that these considerations may be outdated by now.

Long story short, we have more experience with Cython and don't want to add two big dependencies. But of course, in future we might switch to Numba, if this brings major improvements/ enhancements to GSTools.

Why did you ask? - Are you interested in developing for GSTools or are you having troubles with the Cython part of the package?

**Comment by LSchueler** [comment]
@MuellerSeb I am still figuring out how to set the compiler and the compiler flags in TravisCI. As soon as that is done, we can see how the deployment can be done.

**Comment by LSchueler** [comment]
@MuellerSeb I don't think it's much work to cythonize the rest of the code and you are right, we should restructure the modules a bit. - I'll do that when I find some time.

About the broadcasting getting easier: What I meant was the shapes of the pos. and field arrays going into the RandMeth methods, which have to have rather strange shapes at the moment.

**Comment by Hugovdberg** [comment]
> Why did you ask? - Are you interested in developing for GSTools or are you having troubles with the Cython part of the package?
I might be interested in developing as well, but my major concern was that I get the warning that no Cython functions were imported, on all my computers. But since you say it should work for most systems, I retried installing from github and it appears that does work on my Debian 9 machine, contrary to the pipy version. 
On my windows machine I installed packages like `scipy` through `conda` so I don't need a C compiler, while `numba` compiled packages do work on my Windows machine. I think this will be a concern for many more Windows users, where compiling your own software is much less common than in linux environments.

**Comment by LSchueler** [comment]
Hmm, that's interesting. For Linux, MacOS, and for Windows, the Cython code is precompiled and then sent to PyPI. Thus, an installation via pip should include the compiled C code already.

It would be a great help, if you could show us the output of the pip installation, where you get the warnings, that the Cython functions where not imported. - Maybe you are using Python 3.7? - We just recently added Python3.7 support.

In case you really want to add something to GSTools, we'd be more than glad to discuss things with you. Help is always very welcome!

**Comment by LSchueler** [comment]
@Hugovdberg One more thing: The warnings that no Cython functions are imported, simply mean that the variogram estimation is based on a pure Python implementation. - All functionality is still preserved, the estimation is just slower.

**Comment by Hugovdberg** [comment]
I guess it's the fact I'm on Python 3.7 on all my machines. I understand all functionality is available, but still it would be nice to profit from the optimized code.

For completeness, this is the output from `pip` and the warning that remains in python:

```sh
> pip install git+https://github.com/GeoStat-Framework/GSTools.git#egg=gstools
Collecting gstools from git+https://github.com/GeoStat-Framework/GSTools.git#egg=gstools
  Cloning https://github.com/GeoStat-Framework/GSTools.git to c:\users\bergh1\appdata\local\temp\pip-install-8szq34fp\gstools
Requirement already satisfied: numpy>=1.13.0 in c:\anaconda64\envs\cursus_modflow\lib\site-packages (from gstools) (1.16.2)
Requirement already satisfied: scipy>=0.19.1 in c:\anaconda64\envs\cursus_modflow\lib\site-packages (from gstools) (1.2.1)
Requirement already satisfied: hankel>=0.3.6 in c:\anaconda64\envs\cursus_modflow\lib\site-packages (from gstools) (0.3.8)
Requirement already satisfied: emcee in c:\anaconda64\envs\cursus_modflow\lib\site-packages (from gstools) (2.2.1)
Requirement already satisfied: pyevtk in c:\anaconda64\envs\cursus_modflow\lib\site-packages (from gstools) (1.1.1)
Requirement already satisfied: six in c:\anaconda64\envs\cursus_modflow\lib\site-packages (from gstools) (1.12.0)
Requirement already satisfied: mpmath>=0.18 in c:\anaconda64\envs\cursus_modflow\lib\site-packages (from hankel>=0.3.6->gstools) (1.1.0)
Building wheels for collected packages: gstools
  Building wheel for gstools (setup.py) ... done
  Stored in directory: C:\Users\bergh1\AppData\Local\Temp\pip-ephem-wheel-cache-jg0tu270\wheels\64\eb\35\3df8181cbad59a53ca41357f0b6afaacf10a586398f914c187
Successfully built gstools
Installing collected packages: gstools
Successfully installed gstools-1.0.2.dev0

> python
Python 3.7.2 (default, Feb 21 2019, 17:35:59) [MSC v.1915 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import gstools
Warning: No Cython functions imported
```

**Comment by LSchueler** [comment]
@Hugovdberg I'm sorry for the late reply.

You are installing the GSTools pacakge directly from source, which means that on your machines, the Cython code has to first be translated to C code, which in turn has to be compiled.

If you install GSTools with the following command, you should get the most recent version prebuilt:

`pip install -i https://test.pypi.org/simple/ gstools==1.0.2.dev3`

Very soon, we are going to upload one of the newest versions of GSTools to the proper PyPI. But until then you should be fine with the test.pypi package.

Numba also requires a compiler, but it's included in the Numba package, which probably is a smart move of them.

**Comment by LSchueler** [comment]
With the commits from 20.5., the Cython dependency is now obligatory. There will be no more Python fallback methods, which makes all the vectorised Numpy broadcasting stuff a lot easier to read and maintain.

I've also cythonized the `IncomprRandMeth` and the variogram estimation.

Furthermore, all the important and time consuming Cython loops (RandMeth, IncomprRandMeth, and variogram estimation) are now parallelised with OpenMP.
For now, we will not deploy the parallelised versions with PyPI, because we cannot assume that all users have OpenMP libraries installed.
We have some ideas of how to circumvent that problem, but for now, it should be fine, if people, who wish to use OpenMP, build the code via pip on their local machines. - It is still very easy to build, it just increases the number of requirements (Cython, C compiler).
